### ISSUE TITLE
Use this plugin with markdown images

### ISSUE THREAD

**Issue by Brixy** [11ty/eleventy-img#90]
Thank you for this plugin.

It would be great if it was possible to use this plugin with markdown images.

```
![Alt text](/img/image.jpg "Optional title text -> besomes image caption")

![Alt text](/img/image.svg "Optional title text -> besomes image caption")
```

Some ideas:

- Only use this plugin if image wider than a defined width (e.g. 500px).
- Use optional markdown title attribute as image caption (`figcaption`).

Thanks!

**Comment by Alexs7zzh** [comment]
All the code is put in this [gist](https://gist.github.com/Alexs7zzh/d92ae991ad05ed585d072074ea527b5c).

### 1. Transform images in markdown

I am using [transforms](https://www.11ty.dev/docs/config/#transforms) to achieve this.

Since `eleventy-img` provides `Image.statsSync` function, as long as you have `null` in the `widths` option of `eleventy-img`, you can get the width of the original image. You can then compare its width to your defined width to decide if you want to transform the image. 

It can cover all images in your website including the ones from markdown files, as long as you set a proper query selector. (i.e., `article img, main img`)

To be noted, I also use the variable `index` to add `loading=lazy` and `decoding=async` attributes to all images except the first image on every page. Please adjust this and the options for `eleventy-img` depending on your use case.

### 2. Use `figcaption`

To use `figcaption`, I wrote a `markdown-it` plugin which is also included in the gist. Originally an image element will be rendered as:

```html
<p>
  <img src="" alt="" title="" />
</p>
```

With this plugin, it will be rendered as:

```html
// if title is not defined
<figure>
  <img src="" alt="" title="" />
</figure>

// if title is defined
<figure>
  <img src="" alt="" title="title" />
  <figcaption>title</figcaption>
</figure>
```

To use this plugin, you can configure `markdown-it` in your `.eleventy.js` as following:

```js
const md = require('markdown-it')()
  .use(require('./figure')) // the plugin in the gist

eleventyConfig.setLibrary('md', md)
```

**Comment by mahnunchik** [comment]
@Alexs7zzh good approach but I don't understand the reason to use 2 stages. All sync/async image transformation can be done in `transform` filter.

**Comment by solution-loisir** [comment]
Hi @Brixy, I know this is a pretty old discussion and you might already have found a solution or maybe it's not relevant anymore. Anyhow, to answer this:
> It would be great if it was possible to use this plugin with markdown images.

I just published a markdown-it plugin that uses eleventy-img to process images. Check it out [here](https://www.npmjs.com/package/markdown-it-eleventy-img). If you end up trying it, let me know what you think.

**Comment by shirooo39** [comment]
> Hi @Brixy, I know this is a pretty old discussion and you might already have found a solution or maybe it's not relevant anymore. Anyhow, to answer this:
> 
> > It would be great if it was possible to use this plugin with markdown images.
> 
> I just published a markdown-it plugin that uses eleventy-img to process images. Check it out [here](https://www.npmjs.com/package/markdown-it-eleventy-img). If you end up trying it, let me know what you think.

this looks really good! thanks!
I'm still so new to Eleventy but I'll give it a try later.

**Comment by solution-loisir** [comment]
Great! Let me know what you think. :smile:

**Comment by zachleat** [comment]
Related: #208

**Comment by zachleat** [comment]
The docs have been updated to recommend #208 moving forward for synchronous use cases like Markdown. I think we can consider this one resolved with Eleventy v3.0.0-canary.5 and Eleventy Image v4.0.1.

https://www.11ty.dev/docs/plugins/image/#synchronous-shortcode

Thanks y’all!

**Comment by stevenvachon** [comment]
For generating `<figure>` with `width` and `height` attributes, you can do this:
```js
import { markdownItImageSize as MarkdownItImageSize } from 'markdown-it-image-size';
import MarkdownItImplicitFigures from 'markdown-it-implicit-figures';

export default config => {
  config.amendLibrary('md', (mdLib) => {
    mdLib.use(MarkdownItImageSize);
    mdLib.use(MarkdownItImplicitFigures);
  });

  return {
    dir: {
      // etc
    },
  };
};
```
